rend3/src/util/buffer_copier.rs

```diff
@@ -0,0 +1,142 @@
+
+        let pipeline_layout = device.create_pipeline_layout(&wgpu::PipelineLayoutDescriptor {
+            label: Some("BufferCopier Pipeline Layout"),
+            bind_group_layouts: &[&bgl],
+            push_constant_ranges: &[],
+        });
+
+        let module = device.create_shader_module(&wgpu::ShaderModuleDescriptor {
+            label: Some("BufferCopier Compute Shader"),
+            source: wgpu::ShaderSource::Wgsl(include_str!("../../shaders/buffer_copier.wgsl").into()),
+        });
+
+        let pipeline = device.create_compute_pipeline(&wgpu::ComputePipelineDescriptor {
+            layout: Some(&pipeline_layout),
+            label: Some("BufferCopier Pipeline"),
+            module: &module,
+            entry_point: "main",
+        });
+
+        BufferCopier { pipeline, bgl }
+    }
+
+    pub fn execute(&self, device: &Device, queue: &Queue, buffers: [&wgpu::Buffer; 8], params: BufferCopierParams) {
+        let mut encoder = device.create_command_encoder(&wgpu::CommandEncoderDescriptor { label: None });
+
+        let buffer_binding = |idx, buffer| BindGroupEntry {
+            binding: idx,
+            resource: wgpu::BindingResource::Buffer(wgpu::BufferBinding {
+                buffer,
+                offset: 0,
+                size: None,
+            }),
+        };
+
+        let params_buffer = device.create_buffer_init(&wgpu::util::BufferInitDescriptor {
+            label: None,
+            contents: bytemuck::cast_slice(&[params]),
+            usage: wgpu::BufferUsages::UNIFORM,
+        });
+
+        let bind_group = device.create_bind_group(&BindGroupDescriptor {
+            label: Some("BufferCopier Bind Group"),
+            layout: &self.bgl,
+            entries: &[
+                buffer_binding(0, buffers[0]),
+                buffer_binding(1, buffers[1]),
+                buffer_binding(2, buffers[2]),
+                buffer_binding(3, buffers[3]),
+                buffer_binding(4, buffers[4]),
+                buffer_binding(5, buffers[5]),
+                buffer_binding(6, buffers[6]),
+                buffer_binding(7, buffers[7]),
+                BindGroupEntry {
+                    binding: 8,
+                    resource: wgpu::BindingResource::Buffer(wgpu::BufferBinding {
+                        buffer: &params_buffer,
+                        offset: 0,
+                        size: None,
+                    }),
+                },
+            ],
+        });
+
+        {
+            let mut cpass = encoder.begin_compute_pass(&wgpu::ComputePassDescriptor {
+                label: Some("BufferCopier Compute Pass"),
+            });
+
+            cpass.set_pipeline(&self.pipeline);
+            cpass.set_bind_group(0, &bind_group, &[]);
+            let num_workgroups = (params.count + (Self::WORKGROUP_SIZE - 1)) / Self::WORKGROUP_SIZE; // Divide rounding up
+            cpass.dispatch(num_workgroups, 1, 1);
+        }
+
+        queue.submit(Some(encoder.finish()));
```

Comment:
We shouldn't submit here -- we should passthrough this encoder: https://github.com/BVE-Reborn/rend3/blob/trunk/rend3/src/renderer/ready.rs#L18-L20 which will automatically get submitted before everything else.